com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs

```diff
@@ -1997,12 +2005,32 @@ internal void HandleIncomingData(ulong clientId, ArraySegment<byte> payload, flo
         /// </summary>
         /// <param name="clientId">The ClientId to disconnect</param>
         public void DisconnectClient(ulong clientId)
+        {
+            DisconnectClient(clientId, null);
+        }
+
+        /// <summary>
+        /// Disconnects the remote client.
+        /// </summary>
+        /// <param name="clientId">The ClientId to disconnect</param>
+        /// <param name="reason">Disconnection reason. If set, client will receive a DisconnectReasonMessage and have the
+        /// reason available in the NetworkManager.DisconnectReason property</param>
+        public void DisconnectClient(ulong clientId, string reason)
         {
             if (!IsServer)
             {
                 throw new NotServerException($"Only server can disconnect remote clients. Please use `{nameof(Shutdown)}()` instead.");
             }
 
+            if (!string.IsNullOrEmpty(reason))
+            {
+                var disconnectReason = new DisconnectReasonMessage();
+                disconnectReason.Reason = reason;
+                SendMessage(ref disconnectReason, NetworkDelivery.Reliable, clientId);
+
+                MessagingSystem.ProcessSendQueues();
+            }
+
```

Comment:
Actually, I think you want to migrate the MessagingSystem.ProcessSendQueues(); below your string check because any other messages pending will not get sent if this method is invoked.  While DisconnectRemoteClient does call the same method, because OnClientDisconnectFromServer is invoked first the queue is cleared first.

Moving MessagingSystem.ProcessSendQueues(); just outside of the brackets just above OnClientDisconnectFromServer will assure all pending messages in the send queue will get sent to the client being disconnected via DisconnectClient (even if no reason is specified).

Reply:
But then, the other `ProcessSendQueues` should be removed. If only I was not scared of some obscure knock-on bug